### PR TITLE
[feat] file imports in exports map

### PR DESCRIPTION
**Pull request by ignatiusmb**
Related to #1891 

We should also be able to import js files without having to specify the extension. Question is, do we still keep the the export with the extension? I don't know any use case out there that needs to explicitly do `import stuff from 'package/module/stuff.js'`, which can simply be `'package/module/stuff'`

Drafting this first to wait for test cases from #2327

### CODE REVIEW

**Comment by ignatiusmb** [comment]
Does that also applies to imports from `node_modules`? I only encounter those errors from relative paths so far like `./foo/hello.js` or `../utils/index.js`, don't know if `my-module/utils/stuff.js` can be `my-module/utils/stuff` or needs it's extension too.

We can probably export both to be on the safe side? At the cost of increased package.json file size
```json
"exports": {
  "./foo/utils.js": "./foo/utils.js",
  "./foo/utils": "./foo/utils.js"
}
```

**Comment by dummdidumm** [comment]
I think with ESM type imports you need that explicit exports map to import anything not imported through the base path (`package/subpath` vs `package`). But since, with ESM, you need to fully specify the path file, you have to do `import { x } from './foo/index.js'` inside your code, and could have a file named `./foo.js` next to the `./foo` folder.

```
foo.js
foo
  index.js
```

With the export map simplifying `foo.js` to foo, and with the "simplify `foo/index.js` to `foo` logic already present, this would clash. 

That said, I'm not sure if this would ever happen. And it would be highly confusing from a package consumer perspective to have a `foo.js` _and_ a `foo` import present, meaning different things.

**Comment by ignatiusmb** [comment]
Oh, I totally forgot exports field will take higher priority for module resolutions rather than the actual relative paths 😄 

You're also totally right about the second point, this change would always overwrite the `foo/index.js` to `foo.js` if they both exists. Do we want then want `package` to also handle these kinds of overlaps and throw an error, simply prints a warning, or mention it in the docs? When do we draw the line for stuff that is out of `package` command scope? I'm leaning to just mentioning them in the docs for now.

**Review by dummdidumm — APPROVED** [review]
I'm good with these changes. The DX of importing these packages without `.js` surpasses the very small likelyhood of a clash.

**Comment by ignatiusmb** [comment]
Do you think we're missing any odd characters in filenames with the current regex?

**Comment by benmccann** [comment]
I think that for ESM that needing to import with the extension does not apply for things listed in `exports` in `package.json`

**Comment by Rich-Harris** [comment]
Agree that the likelihood of collision is very small, but I do think we should error in that case — it's more likely to be an oversight than anything (an incomplete refactor or whatever) and an error would be preferable to an export no longer working

File: packages/kit/src/packaging/index.js [hunk 1]
```diff
@@ -85,7 +88,15 @@ export async function make_package(config, cwd = process.cwd()) {
 		if (exports_filter(file)) {
 			const entry = `./${out_file.replace(/\\/g, '/')}`;
 			const key = entry.replace(/\/index\.js$|(\/[^/]+)\.js$/, '$1');
+
+			if (clashes[key]) {
```
Comment (ignatiusmb) [inline packages/kit/src/packaging/index.js:92]: I think we can probably reuse `generated` here instead?

**Comment by ignatiusmb** [comment]
Makes sense that an incomplete refactor would produce this, also nice error message! Way better than writing it in the docs

**Comment by ignatiusmb** [comment]
Welp, here we go again. The new tests do fail, but not for reasons I expected.

**Comment by Rich-Harris** [comment]
I don't think we can reuse `generated` — I added `clashes` because otherwise the error message is incorrect in the case where the input file is TypeScript — you should get a message like

> Duplicate "./utils" export. Please remove or rename either $lib/utils/index.ts or $lib/utils.ts

but instead you get 

> Duplicate "./utils" export. Please remove or rename either $lib/utils/index.js or $lib/utils.js

because `generated` contains the output files rather than the input files

**Comment by ignatiusmb** [comment]
Ah, you're right! I didn't take TS files into account, my bad.

**Pull request merged**